src/alchemlyb/preprocessing/subsampling.py

```diff
@@ -34,63 +87,198 @@ def slicing(df, lower=None, upper=None, step=None, force=False):
+                group_c = column.groupby(level=index_names).get_group(name)
+                indices, calc = subsample(group_c, group)
+            elif column in group.columns:
+                group_c = group[column]
+                indices, calc = subsample(group_c, group)
+        elif how is not None:
+            if (how == 'right') or (how == 'left'):
+                try:
+                    group_c = _how_lr(name, group, how)
+                except KeyError:
+                    indices, calc = random_selection(name, group)
+                else:
+                    indices, calc = subsample(group_c, group)
+            elif how == 'random':
+                indices, calc = random_selection(name, group)
+            elif how == 'sum':
+                group_c = _how_sum(name, group)
+                indices, calc = subsample(group_c, group)
+            else:
+                raise ValueError("`how` cannot be '{}';"
+                " see docstring for available options".format(how))
+
+        group_s = slicing(group, lower=lower, upper=upper, step=step)
+        resdata.append(group_s.iloc[indices])
+
+        if return_calculated:
+            calculated[name] = calc
+    
+    if return_calculated:
+        return pd.concat(resdata), pd.DataFrame(calculated)
+    else:
+        return pd.concat(resdata)
+
+
+def statistical_inefficiency(data, column=None, how=None, lower=None,
+        upper=None, step=None, conservative=True, return_calculated=False,
+        force=False, random_state=None):
+    """Subsample an alchemlyb DataFrame based on the calculated statistical
+    inefficiency of one of its columns.
+
+    Calculation of statistical inefficiency and subsequent subsampling will be
+    performed separately on groups of rows corresponding to each set of lambda
+    values present in the DataFrame's index. Each group will be sorted on the
+    outermost (time) index prior to any calculation.
+
+    The `how` parameter determines the observable used for calculating the
+    correlations within each group of samples. The options are as follows:
+
+        'right'
+            The recommended setting for 'u_nk' datasets; the column immediately
+            to the right of the column corresponding to the group's lambda
+            index value is used. If there is no column to the right, then the
+            column to the left is used.  If there is no column corresponding to
+            the group's lambda index value, then 'random' is used (see below).
+        'left'
+            The opposite of the 'right' approach; the column immediately to the
+            left of the column corresponding to the group's lambda index value
+            is used. If there is no column to the left, then the column to the
+            right is used.  If there is no column corresponding to the group's
+            lambda index value, then 'random' is used for that group (see below).
+        'random'
```

Comment:
Why this option?? This seems counterproductive to reproducible research. Does the user even know which column worked in the end? I'd much rather fail.

Reply:
I really do not like this option in a library that tries to be clean and basic. Let the users write their workflows and heuristics. 👎

Reply:
I agree that we would like to avoid randomness where possible. We could remove the `how='random'` parameter choice here, but we may still need to internally address the (admittedly rare?) corner case of a `u_nk` column not existing for a lambda state actually sampled. For the "right"/"left" approaches, if a corresponding column doesn't exist for the lambda state in question, the concept of right and left becomes more guesswork.

Some alternative approaches I can think of:
1. Find the "closest" column in terms of euclidean distance in lambdas, and then go to the right or left of that, but this logic seems to be pretty tortured to me, and I'm not sure if it too can fail in weird ways.
2. Choose the leftmost or rightmost column in the whole DataFrame, and march through columns in order if we hit failures to decorrelate.
3. Try to use the last column we subsampled on (previous iteration of [this loop](https://github.com/alchemistry/alchemlyb/pull/98/files/dc1b3ff106eb1d66f1b13555c6d1abe3d65c0e8d#diff-29b9db00345fef71312e85c1392b6fbeR157)).
4. Fail loudly, and declare this corner case not one we wish to address. The user could get around this by throwing out rows corresponding to lambda values there are no corresponding columns for.

Reply:
One way we can address some of the harder-to-pin-down choices made in alchemlyb is to use a logger (#34). I recognize that this doesn't excuse *strange* choices on the library's part, but if we have a logger in place then at least what the library did in a particular instance is auditable later.

Reply:
Logging would be a good step towards more transparency.

Maybe put in a commented-out logger call with a hint towards #34?

Reply:
Will do! I consider this an important component of the library.